### PR TITLE
GH-82 TransportContext: merge and key iterator

### PR DESCRIPTION
**Pull request by szysas**

### CODE REVIEW

**Comment by sbernard31** [comment]
It's take me few minutes to get how override works. :thinking:   
Actually, when you merge if there is 2 value with same key both are kept  in the chained list but only the new one will be returned by the `get()`, correct ? 

That's surprising me but if it works why not. I guess one drawback is that merged context takes probably a bit more memory than necessary.

I didn't test it but reading the code I guess that : 
```java
TransportContext ctx1 = TransportContext.of(DUMMY_KEY, "111").with(DUMMY_KEY2, "222");
TransportContext ctx2 = TransportContext.of(DUMMY_KEY, "aaa");
TransportContext ctx1Ctx2 = ctx1.with(ctx2);

TransportContext ctx3 = TransportContext.of(DUMMY_KEY, "aaa").with(DUMMY_KEY2, "222");   
// OR TransportContext ctx3 = TransportContext.of(DUMMY_KEY2, "222").with(DUMMY_KEY, "aaa");  
// not sure about the merge order : this makes me think 
// that maybe a better name would be addFirst() or addLast() (instead of with())
// Just because order seems relevant for equals()

assertEquals(ctx1Ctx2, ctx3); // From user point of view, I guess this should be equal, but is it ? reading the code I'm not sure. 
```

and also if I iterate over ctx1Ctx2 and print all key, I will get DUMMY_KEY twice correct ? 

(I maybe totally missed something, sorry I didn't take time to execute code)

**Comment by szysas** [comment]
> and also if I iterate over ctx1Ctx2 and print all key, I will get DUMMY_KEY twice correct ?

Yes, that is correct.

**Comment by szysas** [comment]
> and also if I iterate over ctx1Ctx2 and print all key, I will get DUMMY_KEY twice correct ?

@sbernard31 Now that it is fixed by returning Set with keys.

**Comment by sbernard31** [comment]
I looked at this.
The keys set resolve the "twice" error.

But not the "equals" one ? See code comment in code block of https://github.com/open-coap/java-coap/pull/85#issuecomment-2072391807.

**Comment by szysas** [comment]
> I looked at this. The keys set resolve the "twice" error.
I'm not sure what do you mean.


> But not the "equals" one ? See code comment in code block of [#85 (comment)](https://github.com/open-coap/java-coap/pull/85#issuecomment-2072391807).

Well, that is true, equals will not be true.

**Comment by sbernard31** [comment]
(let me know if I should review this again ?)

**Comment by szysas** [comment]
> (let me know if I should review this again ?)

Yes, please ;)

File: coap-core/src/main/java/com/mbed/coap/transport/TransportContext.java [hunk 1]
```diff
@@ -66,6 +68,17 @@ public <T> TransportContext with(Key<T> key, T value) {
         return new TransportContext(requireNonNull(key), requireNonNull(value), this);
     }
 
+    public TransportContext with(TransportContext otherCtx) {
```
Comment (sbernard31) [inline coap-core/src/main/java/com/mbed/coap/transport/TransportContext.java:71]: Maybe we should add some javadoc to explain that :
If a key exists in both contexts, the `otherCtx` value will be used.

**Comment by sbernard31** [comment]
I'm still a bit confused by the idea that a not used key/value pair should still be stored in memory :sweat_smile:  (https://github.com/open-coap/java-coap/pull/85#issuecomment-2072391807) 

> Actually, when you merge if there is 2 value with same key both are kept in the chained list but only the first one will be returned by the get(), correct ?
> That's surprising me but if it works why not. I guess one drawback is that merged context takes probably a bit more memory than necessary.

But alternative will probably lead to more costly(and complex) way to merge 2 `TransportContext`, so maybe this is the good way. :shrug: 

So expect the optional javadoc addition this looks good to me.

**Comment by szysas** [comment]
Purpose for `TransportContext` is to have short living, simple and small object. When we will see that this is not enough then alternative would be to use some HashMap underneath.

**Pull request merged**